Add ESModule URL + include in errors

rhuanjl opened the pull request:
This is an attempt to implement my own suggestions from: #4068

(Closed first version by mistake)

Allows a host to set a URL for a module with JsSetModuleHostInfo.
Adds this URL to the error stack trace for run time errors
Adds this URL to syntax errors
Doesn't yet include a URL for exports that cant be found, couldn't see how to get the URL here:

https://github.com/Microsoft/ChakraCore/blob/master/lib/Runtime/Language/SourceTextModuleRecord.cpp#L627-L632

boingoing commented on bin/ch/WScriptJsrt.cpp:
```diff
@@ -275,17 +276,21 @@ JsValueRef WScriptJsrt::LoadScriptHelper(JsValueRef callee, bool isConstructCall
         }
 
-        if (!fileNameNarrow && isSourceModule)
+        if(!fileNameNarrow)
```
Style nitpick: we typically have a space between `if` and `(`.

rhuanjl replied:
Done - sorry I should have noticed this from reading your code.

boingoing commented on bin/ch/WScriptJsrt.cpp:
```diff
@@ -390,6 +395,14 @@ JsErrorCode WScriptJsrt::LoadModuleFromString(LPCSTR fileName, LPCSTR fileConten
     // ParseModuleSource is sync, while additional fetch & evaluation are async.
     unsigned int fileContentLength = (fileContent == nullptr) ? 0 : (unsigned int)strlen(fileContent);
+ 
+    if(isFile && fullName)
```
Same note as above.

rhuanjl replied:
done

boingoing commented on bin/ch/WScriptJsrt.cpp:
```diff
@@ -390,6 +395,14 @@ JsErrorCode WScriptJsrt::LoadModuleFromString(LPCSTR fileName, LPCSTR fileConten
+    {
+        JsValueRef moduleURL;
+        ChakraRTInterface::JsCreateString(fullName, strlen(fullName), &moduleURL);
+        errorCode = ChakraRTInterface::JsSetModuleHostInfo(requestModule, JsModuleHostInfo_ModuleURL, moduleURL);
```
This errorCode is lost, right? Seems we should check it directly or use `IfJsrtErrorSetGo`.

boingoing commented on lib/Jsrt/Core/JsrtCore.cpp:
```diff
@@ -77,7 +77,18 @@ JsParseModuleSource(
         if (sourceContextInfo == nullptr)
         {
-            sourceContextInfo = scriptContext->CreateSourceContextInfo(sourceContext, nullptr, 0, nullptr, nullptr, 0);
+            if(moduleRecord->GetModuleURL() != nullptr)
```
Same note above

boingoing commented on lib/Jsrt/Core/JsrtCore.cpp:
```diff
@@ -77,7 +77,18 @@ JsParseModuleSource(
-            sourceContextInfo = scriptContext->CreateSourceContextInfo(sourceContext, nullptr, 0, nullptr, nullptr, 0);
+            if(moduleRecord->GetModuleURL() != nullptr)
+            {
+                Js::JavascriptString *moduleURL = Js::JavascriptString::FromVar(moduleRecord->GetModuleURL());
```
I think I would prefer if this block read as:

```c++
const char16 *moduleURLSz = nullptr;
size_t moduleURLLen = 0;
if (moduleRecord->GetModuleURL())
{
  Js::JavascriptString *moduleURL = Js::JavascriptString::FromVar(moduleRecord->GetModuleURL());
  moduleURLSz = moduleURL->GetSz();
  moduleURL->GetLength();
}
sourceContextInfo = scriptContext->CreateSourceContextInfo(sourceContext, moduleURLSz, moduleURLLen, nullptr, nullptr, 0);
```

Maybe just a personal preference, but this is simpler for me to understand.

boingoing commented on lib/Runtime/Language/SourceTextModuleRecord.cpp:
```diff
@@ -171,7 +172,14 @@ namespace Js
             if (*exceptionVar == nullptr)
             {
-                *exceptionVar = JavascriptError::CreateFromCompileScriptException(scriptContext, &se);
+                if(this->GetModuleURL() != nullptr)
```
Same note as above

boingoing commented on lib/Runtime/Language/SourceTextModuleRecord.h:
```diff
@@ -50,6 +50,10 @@ namespace Js
         Var GetSpecifier() const { return normalizedSpecifier; }
         const char16 *GetSpecifierSz() const { return JavascriptString::FromVar(this->normalizedSpecifier)->GetSz(); }
 
+        void SetModuleURL(Var moduleURL) {this->moduleURL = moduleURL;}
```
These small in-line methods like this should have spaces padding the body from the curly-braces. (`{this}` => `{ this }`)

boingoing commented on lib/Runtime/Language/SourceTextModuleRecord.cpp:
```diff
@@ -831,7 +839,14 @@ namespace Js
         if (rootFunction == nullptr)
         {
-            this->errorObject = JavascriptError::CreateFromCompileScriptException(scriptContext, &se);
+            if(this->GetModuleURL() != nullptr)
```
One more

boingoing commented on lib/Runtime/Language/SourceTextModuleRecord.cpp:
```diff
@@ -171,7 +172,14 @@ namespace Js
+                }
+                else
+                {
+                    *exceptionVar = JavascriptError::CreateFromCompileScriptException(scriptContext, &se);                    
```
I would prefer if this and the block below fetched the source URL before calling `JavascriptError::CreateFromCompileScriptException` so that we could make this a bit simpler. More like:

```c++
const WCHAR * sourceURL = nullptr;
if (this->GetModuleURL())
{
  sourceURL = this->GetModuleURLSz();
}
*exceptionVar = JavascriptError::CreateFromCompileScriptException(scriptContext, &se, sourceURL);
```

rhuanjl commented:
Thanks for the review @boingoing I've addressed the style points, thinking about how to add a test for this, will try and do shortly.

fatcerberus commented:
Should `moduleURL` be `moduleUrl`?  Seems to be the convention throughout the rest of the codebase.

rhuanjl commented:
@boingoing I've added a test as well as fixing the style points and I did the rename that @fatcerberus suggested as it seemed sensible to me. I've rebased and squashed to make it clean. Please let me know if I should do anything else.

boingoing commented:
@rhuanjl Would you mind rebasing your change on top of master? It has gotten out-of-date and I would prefer to let you do the rebase. Please keep the squash commit, though. Thank you. 😃

rhuanjl commented:
@boingoing I rebased and it got out of date again so I rebased again... Hopefully it can go in before it needs rebasing again, let me know if you need anything else from me.

Thanks agin for the review.

rhuanjl commented:
@boingoing I rebased again, please do let me know if you need me to do anything else.

boingoing commented:
@rhuanjl Sorry for delays on this one. A piece of our internal infrastructure was broken blocking me from pulling-in your change.

rhuanjl commented:
@boingoing Thanks for getting it sorted.
